geode-core/src/main/java/org/apache/geode/cache/client/internal/ServerRegionProxy.java

```diff
@@ -756,13 +756,14 @@ public void executeFunction(String functionId,
         serverRegionExecutor, resultCollector, hasResult, isHA, optimizeForWrite);
 
     int retryAttempts = pool.getRetryAttempts();
+    boolean inTransaction = TXManagerImpl.getCurrentTXState() != null;
 
     final Supplier<AbstractOp> executeRegionFunctionOpSupplier =
         () -> new ExecuteRegionFunctionOp.ExecuteRegionFunctionOpImpl(region.getFullPath(),
             functionId, serverRegionExecutor, resultCollector, hasResult, isHA,
             optimizeForWrite, true, timeoutMs);
 
-    if (pool.getPRSingleHopEnabled()) {
+    if (pool.getPRSingleHopEnabled() && !inTransaction) {
```

Comment:
This basically avoids using singleHop for transactional operation; the reason for that is, the single hop ops are performed in executor thread, losing the transaction context. Options could be to see, if tx context can be passed to the singlehop executors or to move the logic from SingleHopOperationCallable to separate method and call it in the same execution thread. This could be done as part of separate ticket/issue.

Reply:
Yeah, changing the logic should be a different ticket. This ticket is concerned with making the two code paths consistent. The logic came from https://issues.apache.org/jira/browse/GEODE-6016.